File: Ontology/DataIntersectionOf.cs
using System.Collections.Generic;
using System.Linq;

namespace Ontology
{
    public class DataIntersectionOf: IDataIntersectionOf
    {
        private readonly IList<IDataRange> _dataRanges;

        public DataIntersectionOf(
            params IDataRange[] dataRanges
            )
        {
            _dataRanges = dataRanges;
        }

        IList<IDataRange> IDataIntersectionOf.DataRanges => _dataRanges;

        bool IDataRange.HasMember(
            object value
            ) => _dataRanges.All(dataRange => dataRange.HasMember(value));
    }
}

File: Web/ClientApp/src/app/Ontology/DataIntersectionOf.ts
import { DataRange } from "./DataRange";
import { IDataIntersectionOf } from "./IDataIntersectionOf";
import { IDataRange } from "./IDataRange";
import { IDataRangeSelector } from "./IDataRangeSelector";

export class DataIntersectionOf
    extends DataRange
    implements IDataIntersectionOf
{
    constructor(
        public DataRanges: IDataRange[]
        )
    {
        super();
    }

    HasMember(
        value: any
        ): boolean
    {
        return this.DataRanges.every(dataRange => dataRange.HasMember(value));
    }

    Select<TResult>(
        selector: IDataRangeSelector<TResult>
        ): TResult
    {
        return selector.DataIntersectionOf(this);
    }
}
